{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "26d805f7",
   "metadata": {},
   "source": [
    "# Clustering\n",
    "## Input\n",
    "We will cluster the sequences using MMseqs2. First we need to combine the sequences we downloaded into one file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "768f62a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from Bio import SeqIO\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "187d8b01",
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_file = \"all_proteins_combined.fasta\"\n",
    "data_dir = \"../data\"\n",
    "cluster_file = \"mmseqs_output_cluster.tsv\"\n",
    "path_raw_gene_families = \"../raw_gene_families\"\n",
    "path_clean_families = \"../filtered_gene_families\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b80609da",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating all_proteins_combined.fasta...\n",
      "Merged 55 genomes into all_proteins_combined.fasta\n"
     ]
    }
   ],
   "source": [
    "#combine all fasta files for mmseqs\n",
    "\n",
    "\n",
    "print(f\"Creating {seq_file}...\")\n",
    "\n",
    "with open(seq_file, \"w\") as outfile:\n",
    "    count = 0\n",
    "    for filename in os.listdir(data_dir):\n",
    "        if filename.endswith(\".faa\"):\n",
    "            #use filename as species label (like Bifidobacterium_longum)\n",
    "            species_tag = filename.replace(\".faa\", \"\")\n",
    "            \n",
    "            with open(os.path.join(data_dir, filename), \"r\") as infile:\n",
    "                for line in infile:\n",
    "                    if line.startswith(\">\"):\n",
    "                        \n",
    "                        original_id = line.split()[0].lstrip(\">\")\n",
    "                        outfile.write(f\">{species_tag}|{original_id}\\n\")\n",
    "                    else:\n",
    "                        outfile.write(line)\n",
    "            count += 1\n",
    "\n",
    "print(f\"Merged {count} genomes into {seq_file}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f8d001f",
   "metadata": {},
   "source": [
    "## Running clustering script\n",
    "Now we run MMseqs2 on all_proteins_combined.fasta using *run_mmseqs.py*. I decided to do this via a separate script so that I can run it on the Entropy compute cluster and make it faster.\n",
    "\n",
    "On the cluster I simply open a bash session\n",
    "\n",
    "*srun --pty --cpus-per-task=4 --time=100 bash*\n",
    "\n",
    "and run the script\n",
    "\n",
    "*time python3 run_mmseqs.py*\n",
    "\n",
    "which utilzed only took 1min 11s, thanks to utilizing 40 CPU cores.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90548ce6",
   "metadata": {},
   "source": [
    "## Output\n",
    "We obtain 3 files:\n",
    "* mmseqs_output_cluster.tsv - this file says which sequence belongs to which cluster. Contains two columns: cluster name (a cluster is named after one of its representatives) and name of the sequence belonging to it. \n",
    "* mmseqs_output_rep_seq.fasta - those representative sequences clusters are named after \n",
    "* mmseqs_output_all_seqs.fasta - all the sequences reorganized in accordance with the clustering.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6dd17cd9",
   "metadata": {},
   "source": [
    "### Gene families"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1840c1d1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28757 clusters in total\n"
     ]
    }
   ],
   "source": [
    "\n",
    "#load clusters\n",
    "seq_dict = SeqIO.to_dict(SeqIO.parse(seq_file, \"fasta\"))\n",
    "clusters = defaultdict(list)\n",
    "with open(cluster_file) as f:\n",
    "    for line in f:\n",
    "        rep, member = line.strip().split('\\t')\n",
    "        clusters[rep].append(member)\n",
    "\n",
    "n_clusters = len(clusters)\n",
    "print(f\"{n_clusters} clusters in total\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c6b4fc00",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0/28757\n",
      "1000/28757\n",
      "2000/28757\n",
      "3000/28757\n",
      "4000/28757\n",
      "5000/28757\n",
      "6000/28757\n",
      "7000/28757\n",
      "8000/28757\n",
      "9000/28757\n",
      "10000/28757\n",
      "11000/28757\n",
      "12000/28757\n",
      "13000/28757\n",
      "14000/28757\n",
      "15000/28757\n",
      "16000/28757\n",
      "17000/28757\n",
      "18000/28757\n",
      "19000/28757\n",
      "20000/28757\n",
      "21000/28757\n",
      "22000/28757\n",
      "23000/28757\n",
      "24000/28757\n",
      "25000/28757\n",
      "26000/28757\n",
      "27000/28757\n",
      "28000/28757\n"
     ]
    }
   ],
   "source": [
    "#write them to indivitual files\n",
    "\n",
    "for i, (rep, members) in enumerate(clusters.items()):\n",
    "    if not i%1000:\n",
    "        print(f\"{i}/{n_clusters}\")\n",
    "\n",
    "    records = [seq_dict[m] for m in members if m in seq_dict]\n",
    "    \n",
    "    #replace special characters for valid file name\n",
    "    safe_name = rep.replace(\"|\", \"_\").replace(\"/\", \"_\")\n",
    "    output_path = os.path.join(path_raw_gene_families, f\"cluster_{safe_name}.fasta\")\n",
    "    os.makedirs(os.path.dirname(output_path), exist_ok=True)\n",
    "    SeqIO.write(records, output_path, \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "343ae4f5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0/28757\n",
      "1000/28757\n",
      "2000/28757\n",
      "3000/28757\n",
      "4000/28757\n",
      "5000/28757\n",
      "6000/28757\n",
      "7000/28757\n",
      "8000/28757\n",
      "9000/28757\n",
      "10000/28757\n",
      "11000/28757\n",
      "12000/28757\n",
      "13000/28757\n",
      "14000/28757\n",
      "15000/28757\n",
      "16000/28757\n",
      "17000/28757\n",
      "18000/28757\n",
      "19000/28757\n",
      "20000/28757\n",
      "21000/28757\n",
      "22000/28757\n",
      "23000/28757\n",
      "24000/28757\n",
      "25000/28757\n",
      "26000/28757\n",
      "27000/28757\n",
      "28000/28757\n"
     ]
    }
   ],
   "source": [
    "#filtering paralogs and renaming the sequences to contain only names of the species\n",
    "os.makedirs(path_clean_families, exist_ok=True)\n",
    "os.makedirs(path_raw_gene_families, exist_ok=True)\n",
    "\n",
    "n = len(clusters)\n",
    "\n",
    "for i, (rep, members) in enumerate(clusters.items()):\n",
    "    if not i % 1000:\n",
    "        print(f\"{i}/{n}\")\n",
    "\n",
    "    species_best = {}\n",
    "    all_records = []\n",
    "\n",
    "    for member in members:\n",
    "        if member not in seq_dict: continue\n",
    "        \n",
    "        # parse id: Species|ProteinID\n",
    "        parts = member.split('|')\n",
    "        species = parts[0]\n",
    "        pid = parts[1] if len(parts) > 1 else \"unk\"\n",
    "        \n",
    "        record = seq_dict[member]\n",
    "\n",
    "        # 1. filter best (longest)\n",
    "        if species not in species_best or len(record.seq) > len(species_best[species].seq):\n",
    "            species_best[species] = record\n",
    "\n",
    "        # 2. rename paralogs (unique headers required)\n",
    "        raw_rec = record[:] \n",
    "        raw_rec.id = f\"{species}_{pid}\" \n",
    "        raw_rec.description = \"\"\n",
    "        all_records.append(raw_rec)\n",
    "\n",
    "    # save clean (1-to-1)\n",
    "    clean_records = []\n",
    "    for species, record in species_best.items():\n",
    "        new_rec = record[:] \n",
    "        new_rec.id = species\n",
    "        new_rec.description = \"\" \n",
    "        clean_records.append(new_rec)\n",
    "\n",
    "    safe_name = rep.replace(\"|\", \"_\").replace(\"/\", \"_\")\n",
    "    \n",
    "    if clean_records:\n",
    "        SeqIO.write(clean_records, os.path.join(path_clean_families, f\"cluster_{safe_name}.fasta\"), \"fasta\")\n",
    "\n",
    "    if all_records:\n",
    "        SeqIO.write(all_records, os.path.join(path_raw_gene_families, f\"cluster_{safe_name}.fasta\"), \"fasta\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8cfe489",
   "metadata": {},
   "source": [
    "# Alignments\n",
    "Now we will align the sequences using MAFFT. Both for the gene families with paralogs removed (*filtrered_gene_families*) and ones with paralogs present (*raw_gena_families*), in case we wanna use them later. Took 27min 16s.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78d44394",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "05bbb5e2",
   "metadata": {},
   "source": [
    "# Gene trees\n",
    "Computing gene trees with FastTree took 5min 39s. Computations were run togther for filtered_gene_families (no paralogs) and raw_gene_families (with paralogs), because if there's less then 4 trees in the filtered version of a gene family, we remove the unfiltered version too, since it contains less than 4 unique species."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (Enformer venv)",
   "language": "python",
   "name": "enformer_pytorch_env_311"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
